joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+                                      automatically_declare_parameters_from_overrides=True)
+
+        self.declare_ros_parameter('publish_default_efforts', False, ParameterDescriptor(type=ParameterType.PARAMETER_BOOL))
+        self.declare_ros_parameter('publish_default_positions', True, ParameterDescriptor(type=ParameterType.PARAMETER_BOOL))
+        self.declare_ros_parameter('publish_default_velocities', False, ParameterDescriptor(type=ParameterType.PARAMETER_BOOL))
+        self.declare_ros_parameter('rate', 10, ParameterDescriptor(type=ParameterType.PARAMETER_INTEGER))
+        self.declare_ros_parameter('source_list', [], ParameterDescriptor(type=ParameterType.PARAMETER_STRING_ARRAY))
+        self.declare_ros_parameter('use_mimic_tags', True, ParameterDescriptor(type=ParameterType.PARAMETER_BOOL))
+        self.declare_ros_parameter('use_smallest_joint_limits', True, ParameterDescriptor(type=ParameterType.PARAMETER_BOOL))
+        # In theory we would also declare 'dependent_joints' and 'zeros' here.
+        # Since rclpy doesn't support maps natively, though, we just end up
+        # letting 'automatically_declare_parameters_from_overrides' declare
+        # any parameters for us.
+
+        self.free_joints = {}
+        self.joint_list = [] # for maintaining the original order of the joints
+        self.dependent_joints = self.parse_dependent_joints()
+        self.use_mimic = self.get_param('use_mimic_tags')
+        self.use_small = self.get_param('use_smallest_joint_limits')
+
+        zeros = self.node.get_parameters_by_prefix('zeros')
+        # get_parameters_by_prefix() returns a map of name -> Parameter
+        # structures, but self.zeros is expected to be a list of name -> float;
+        # fix that here.
+        self.zeros = {k:v.value for (k, v) in zeros.items()}
+
+        self.pub_def_positions = self.get_param('publish_default_positions')
+        self.pub_def_vels = self.get_param('publish_default_velocities')
+        self.pub_def_efforts = self.get_param('publish_default_efforts')
+
+        self.running = True  # Will be set to False by the Qt window to quit the main loop
+
+        if urdf_file is not None:
+            # If we were given a URDF file on the command-line, use that.
+            with open(urdf_file, 'r') as infp:
+                description = infp.read()
+            self.configure_robot(description)
+        else:
+            # Otherwise, subscribe to the '/robot_description' topic and wait
+            # for a callback there
+            self.node.get_logger().info('Waiting for robot_description to be published on the robot_description topic...')
+            self.node.create_subscription(std_msgs.msg.String, 'robot_description',
+                                          lambda msg: self.configure_robot(msg.data),
+                                          rclpy.qos.QoSProfile(depth=10, durability=rclpy.qos.QoSDurabilityPolicy.RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL))
```

Comment:
nit depth of 1 seems fine. If multiple `robot_description` updates come in a row, I don't think anything is lost as long as the very last one gets acted on.

Reply:
Fixed in 2917803.